src/main/scala/tech/sourced/gemini/ConnectedComponents.scala

```diff
@@ -0,0 +1,125 @@
+package tech.sourced.gemini
+
+import java.nio.ByteBuffer
+
+import com.datastax.driver.core.{Session, SimpleStatement}
+import org.slf4j.{Logger => Slf4jLogger}
+import scala.collection.JavaConverters._
+import scala.collection.mutable.ListBuffer
+
+case class FileHash(sha1: String, value: ByteBuffer)
+
+class DBConnectedComponents(
+                              log: Slf4jLogger,
+                              conn: Session,
+                              keyspace: String = Gemini.defautKeyspace) extends ConnectedComponents(log) {
+  def getHashtables(): List[Byte] = {
+    val hashtablesCql = s"SELECT DISTINCT hashtable FROM $keyspace.hashtables"
+
+    conn
+      .execute(new SimpleStatement(hashtablesCql))
+      .asScala
+      .map(_.getByte("hashtable"))
+      .toList.
+      sorted
+  }
+
+  def getHashValues(hashtable: Byte): Iterable[FileHash] = {
+    val cql = s"SELECT sha1, value FROM $keyspace.hashtables WHERE hashtable=$hashtable"
+
+    conn.execute(new SimpleStatement(cql))
+      .asScala
+      .map(row => FileHash(row.getString("sha1"), row.getBytes("value")))
+  }
+}
+
+abstract class ConnectedComponents(log: Slf4jLogger) {
+  def getHashtables(): List[Byte]
+
+  def getHashValues(hashtable: Byte): Iterable[FileHash]
+
+  def makeBuckets(): List[List[Int]] = {
+    val (buckets, elementIds) = getHashtables()
+      .foldLeft(List[List[Int]](), collection.mutable.Map[String, Int]()) {
```

Comment:
If it's important to have a mutable version here for some reason, here and everywhere else it's better to follow http://twitter.github.io/effectivescala/#Formatting-Imports

Reply:
this is the only place where I use mutable structure. Imo, it fits here. It's possible to rewrite using the immutable map but it would complicate code without any advantages. Usage of a mutable map is scoped to only this function.
It works similar to cache and using mutable map for cache is recommended in https://docs.scala-lang.org/overviews/collections/maps.html